examples/bq-email-exports/main.py

```diff
@@ -0,0 +1,181 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# using SendGrid's Python Library
+# https://github.com/sendgrid/sendgrid-python
+
+import os
+import time
+import datetime
+from sendgrid import SendGridAPIClient
+from sendgrid.helpers.mail import Mail
+from google.cloud import bigquery, storage
+from google.auth import iam, default
+from google.auth.transport import requests
+from google.oauth2 import service_account
+
+
+def credentials():
+    """Gets credentials to authenticate Google APIs.
+
+    Args:
+        None
+
+    Returns:
+        Credentials to authenticate the API.
+    """
+    # Get Application Default Credentials if running in Cloud Functions
+    if os.getenv("IS_LOCAL") is None:
+        credentials, project = default(
+            scopes=["https://www.googleapis.com/auth/cloud-platform"])
+    # To use this file locally set IS_LOCAL=1 and populate env var GOOGLE_APPLICATION_CREDENTIALS
+    # with path to service account json key file
+    else:
+        credentials = service_account.Credentials.from_service_account_file(
+            os.getenv("GOOGLE_APPLICATION_CREDENTIALS"),
+            scopes=["https://www.googleapis.com/auth/cloud-platform"],
+        )
+    return credentials
+
+
+def main():
+    """Sends an email with a Google Cloud Storage signed URL of BQ Query results.
+    Creates BQ table, runs a SQL query, and exports the results to Cloud Storage as a CSV.
+    Generates signing credentials for the CSV and sends an email with a link to the signed URL.
+
+    Args:
+        None
+
+    Returns:
+        None
+    """
+
+    # Create BQ and Storage Client
+    bq_client = bigquery.Client(credentials=credentials())
+    storage_client = storage.Client(credentials=credentials())
```

Comment:
You don't need to create your own credentials object. By default calling "bigquery.Client()" will pick up the environment credentials. If you're running it local, it too will pick up your local environment credentials.

BigQuery example: https://googleapis.dev/python/bigquery/latest/index.html#example-usage
Cloud Function example usage of pub/sub client: https://github.com/GoogleCloudPlatform/python-docs-samples/blob/f6fb9e1c94aa6ffdf4f9460f8c203cb551112438/functions/tips/main.py#L152

Reply:
I may have spoken to soon. I just ran into this statement regarding storage signed urls:

"Note that this method requires a service account key file. You can not use
    this if you are using Application Default Credentials from Google Compute
    Engine or from the Google Cloud SDK."

Reply:
@danieldeleo So the problem with this was we were running into errors with ADC where we had to explicitly give a service account json key for the signing to work if run locally.  When run on GCE it was able to pick up the metadata key and that worked fine.

Reply:
@danieldeleo sorry your second message just loaded :)

Reply:
Actually I think Cloud Functions is not included as part of that statement. Can you please test with my recommendations?

Reply:
It is used while creating iam.Signer() which ultimately gets passed to signed url(). We did try that before but will test with your recommendations again!

Reply:
https://gist.github.com/jezhumble/91051485db4462add82045ef9ac2a0ec

Let's give this a whirl

Reply:
So I think that workaround is for making this work on GCE/GCF but doesn't address the problem with ADC and G Cloud SDK. It also introduces dependency on the GCE metadata server but yeah worth another look.